replication/event.go

```diff
@@ -254,9 +298,14 @@ func (e *PreviousGTIDsEvent) Decode(data []byte) error {
 			}
 			intervals[i] = interval
 		}
-		previousGTIDSets[i] = fmt.Sprintf("%s:%s", uuid, strings.Join(intervals, ":"))
+		if isTag {
+			previousGTIDSets[currentSetnr-1] += fmt.Sprintf(":%s:%s", tag, strings.Join(intervals, ":"))
```

Comment:
Because string is immutable, string `+=` will create a new string object each time. If there are too many tags the performance will drop. Maybe store every "tag+intervals" and join them once at all

Reply:
maybe try this

```diff
diff --git a/replication/event.go b/replication/event.go
index 20c7b0c..0786516 100644
--- a/replication/event.go
+++ b/replication/event.go
@@ -277,7 +277,9 @@ func (e *PreviousGTIDsEvent) Decode(data []byte) error {
 	pos += 8
 
 	previousGTIDSets := make([]string, uuidCount)
+
 	currentSetnr := 0
+	var buf strings.Builder
 	for range previousGTIDSets {
 		uuid := e.decodeUuid(data[pos : pos+16])
 		pos += 16
@@ -292,30 +294,38 @@ func (e *PreviousGTIDsEvent) Decode(data []byte) error {
 				pos += tagLength
 			}
 		}
+
+		if isTag {
+			buf.WriteString(":")
+			buf.WriteString(tag)
+		} else {
+			if currentSetnr != 0 {
+				buf.WriteString(",")
+			}
+			buf.WriteString(uuid)
+			currentSetnr += 1
+		}
+
 		sliceCount := binary.LittleEndian.Uint16(data[pos : pos+8])
 		pos += 8
-		intervals := make([]string, sliceCount)
-		for i := range intervals {
+		for range sliceCount {
+			buf.WriteString(":")
+
 			start := e.decodeInterval(data[pos : pos+8])
 			pos += 8
 			stop := e.decodeInterval(data[pos : pos+8])
 			pos += 8
-			interval := ""
 			if stop == start+1 {
-				interval = fmt.Sprintf("%d", start)
+				fmt.Fprintf(&buf, "%d", start)
 			} else {
-				interval = fmt.Sprintf("%d-%d", start, stop-1)
+				fmt.Fprintf(&buf, "%d-%d", start, stop-1)
 			}
-			intervals[i] = interval
 		}
-		if isTag {
-			previousGTIDSets[currentSetnr-1] += fmt.Sprintf(":%s:%s", tag, strings.Join(intervals, ":"))
-		} else {
-			previousGTIDSets[currentSetnr] = fmt.Sprintf("%s:%s", uuid, strings.Join(intervals, ":"))
+		if !isTag {
 			currentSetnr += 1
 		}
 	}
-	e.GTIDSets = strings.Join(previousGTIDSets[:currentSetnr], ",")
+	e.GTIDSets = buf.String()
 	return nil
 }
 
```